typing_indicators: Add typing indicators for streams.

prakhar1144 opened the pull request:
Rebased #17040 and minor changes.


[typing.webm](https://github.com/zulip/zulip/assets/56781761/b6e4867a-0ac1-4ff6-951b-b066961389aa)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented on web/src/narrow_state.js:
I think we should squash this commit with the next commit ?

timabbott replied:
I think it's OK to have a commit that adds a new function to a shared library like this, as long as it has tests.

prakhar1144 commented on web/src/typing_data.ts:
```diff
@@ -3,11 +3,13 @@ import * as util from "./util";
 
-const typist_dct = new Map<string, number[]>();
+const pm_typists_dict = new Map<string, number[]>();
+const stream_typists_dict = new Map<string, number[]>();
```
Should we rename `stream_typists_dict` to `stream_topic_typists_dict` or more simply `topic_typists_dict`?

Reason: 

Currently our dict is `stream_typists_dict` and we use the term `get_topic_key`.

For PMs, it looks better: `pm_typists_dict` and `get_pms_key`

timabbott replied:
`topic_typists_dict` seems clear enough to me, but I wouldn't object to `stream_topic_` or `stream_conversation_typists_dict`.

timabbott commented on web/shared/src/typing_status.ts:
```diff
@@ -93,23 +90,25 @@ export function maybe_ping_server(
-        // on arrays.
-        if (_.isEqual(new_recipient_ids, state.current_recipient_ids)) {
+        // on arrays and objects.
+        if (_.isEqual(new_recipient, state.current_recipient)) {
```
I'm not sure this check is correct given topics being case-insensitive; we may need to use something like `util.same_recipient` or `util.same_stream_and_topic`.

prakhar1144 replied:
@timabbott 

Actually, `util.ts` has 
```
export type Recipient =
    | {type: "private"; to_user_ids?: string; reply_to: string}
    | ({type: "stream"} & StreamTopic);
```
```
function util_same_recipient(a?: Recipient, b?: Recipient): boolean { }
```
We can use them. They are a fit for our usecase.

One doubt I have is "what is our philosophy behind keeping files in `web/shared/src` and `web/src`? 
`typing_status.ts` and `util.ts` lie in those two different places.

timabbott replied:
So `web/shared/src` is for sharing code with the React Native mobile apps -- it's intended to be self-contained so that it can be used from that other codebase.

Given that we're rewriting the mobile apps in Flutter, it's not a very high priority to maintain that distinction long term. Let me start a discussion on what the plan should be for that.

prakhar1144 replied:
> Let me start a discussion on what the plan should be for that.

@timabbott Did you start a discussion?

gnprice replied:
Good question @prakhar1144. It looks like @timabbott did start a chat discussion on this last week shortly after that comment, but just forgot to add a link to it from here. The discussion is at:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/web.2Fshared.2Fsrc.20in.20zulip.2Fzulip/near/1651184

timabbott replied:
Posted in that thread to move this forward. And yeah, sorry for forgetting to backlink the discussion after starting it.

prakhar1144 replied:
Okay, while working on this, I realized that the functions in `utils.ts` can't be used as it is. So just created a couple of similar functions for our check of `same_recipients`.

timabbott commented on web/src/typing.js:
```diff
@@ -38,6 +36,37 @@ function send_typing_notification_ajax(user_ids_array, operation) {
+        Array.isArray(to) &&
+        message_type === "private" &&
+        user_settings.send_private_typing_notifications
+    ) {
```
Yeah, that check seems a bit odd -- I feel like a `message_type` check should be sufficient to know what type it is.

timabbott commented on docs/subsystems/typing-indicators.md:
```diff
@@ -133,7 +133,7 @@ like the following:
 - `remove_typist`
 - `get_group_typists`
-- `get_all_typists`
+- `get_all_pms_typists`
```
We renamed "private message" to "direct messages", so this name is stale. We've also generally been trying to avoid abbreviations across the codebase; I think we can go with `get_all_direct_message_typists`.

timabbott commented on web/src/typing_data.ts:
```diff
@@ -11,22 +11,20 @@ export function clear_for_testing(): void {
 }
 
-function get_key(group: number[]): string {
+export function get_pms_key(group: number[]): string {
```
Maybe `get_direct_message_conversation_key`?

timabbott commented on web/tests/typing_events.test.js:
```diff
@@ -45,6 +45,7 @@ people.add_active_user(kitty);
 run_test("render_notifications_for_narrow", ({override, mock_template}) => {
     override(page_params, "user_id", anna.user_id);
     const group = [anna.user_id, vronsky.user_id, levin.user_id, kitty.user_id];
+    const pms_typing_key = typing_data.get_pms_key(group);
```
`conversation_key` might be a better name for this local variable in the test.

timabbott commented on web/src/narrow_state.js:
```diff
@@ -126,6 +126,14 @@ export function stream_name() {
+        return undefined;
+    }
+    return stream_data.get_stream_id(stream);
+}
```
I would prefer for this to perhaps call `stream_sub` and access `.id` if present -- we're in the process of migrating away from the `stream_name` being a core part of how we reference data in the codebase, and doing so avoids adding another copy of `stream_name => id` lookup logic.

timabbott commented on zerver/lib/home.py:
```diff
@@ -137,7 +137,7 @@ def build_page_params_for_home_page_load(
         "bulk_message_deletion": True,
         "user_avatar_url_field_optional": True,
-        "stream_typing_notifications": False,  # Set this to True when frontend support is implemented.
+        "stream_typing_notifications": True,
```
Can you start an "api design" conversation posting the current typing notifications API for stream messages just to get eyes on it before we merge this? 

Before we turn this on is a good opportunity to make any changes to improve that API -- E.g., if we hadn't already done it, this would be a good opportunity to rename `private` to `direct` as the message type for these without clients needing to write any compatibility code if we do it before merging the feature.

prakhar1144 replied:
[CZO discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20typing.20indicator/near/1649887)

prakhar1144 replied:
PR: #26958

timabbott commented on web/src/typing_data.ts:
```diff
@@ -16,39 +18,61 @@ export function get_pms_key(group: number[]): string {
+        return pm_typists_dict;
+    }
+    throw new Error(`Unknown message_type: ${message_type}`);
+}
```
I can't tell whether I'm missing something, but do we strictly need two dictionaries? It seems like we could just have keys have the structure `stream:{topic_key}` or `direct:{direct_message_key}`, and cut at least this section of

Not sure, and ultimately this isn't a ton of if/else logic regardless, but figure it's worth mentioning.

timabbott commented:
Thanks @prakhar1144 -- the screencast looks great, but I posted a bunch of comments on the code, mostly about variable naming. I think one other thing to check is whether we're using the modern "direct" name in the API (see the API changes note in https://zulip.com/api/set-typing-status) -- that seems like the kind of opportunity we might miss in having rebased a PR that originated before that.

prakhar1144 commented:
The first 5 commits are from #26858 for manual testing. We can drop them once that PR is merged.

Updates in the recent push:
* Addressed the review comments.
* Changes to use `stream_id` instead of `[stream_id]` and `message_type` as "direct" instead of "private". -- Changes to align with the API changes we made.
* 3rd last commit -- rename `typist_dct` to `typists_dict`
* Added test in the first commit for `stream message` case.

timabbott commented on web/shared/src/typing_status.ts:
```diff
@@ -1,26 +1,64 @@
+    } else if (message_type(a) === "stream" && message_type(b) === "stream") {
+        // type assertions to avoid linter error
+        const aStreamTopic = a as StreamTopic;
+        const bStreamTopic = b as StreamTopic;
```
These can likely be avoided by just asserting they're not integers, given the types.

timabbott commented:
This looks good, marking to merge once CI passes. We should fix https://github.com/zulip/zulip/pull/26904#discussion_r1353681645 in a near term follow-up. @laurynmm it'll probably be worth your doing a polish pass on the API documentation for this; I think the information is all correct though.

@alya can you plan updates to /help/ documentation and the like? We may also want to think about adjustments to the permissions settings around typing notifications.

alya commented:
I filed #27163 so that we don't forget to update the help center. I'll do a /features update if needed when working on 8.0 release prep.
